compiler/rustc_hir_typeck/src/fn_ctxt/adjust_fulfillment_errors.rs

```diff
@@ -67,6 +74,33 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             })
         };
 
+        // Account for enum variant constructors, where the type param corresponds to the enum
+        // itself.
+        let enum_def_id = if let hir::def::DefKind::Ctor(hir::def::CtorOf::Variant, _) =
+            self.tcx.def_kind(def_id)
+        {
+            // `def_id` corresponds to a constructor, and its parent is the variant, and we want
+            // the enum.
+            Some(self.tcx.parent(self.tcx.parent(def_id)))
+        } else {
+            None
+        };
+        let variant_param_to_point_at = find_param_matching(&|param_term| {
+            // FIXME: It would be nice to make this not use string manipulation,
+            // but it's pretty hard to do this, since `ty::ParamTy` is missing
+            // sufficient info to determine if it is synthetic, and we don't
+            // always have a convenient way of getting `ty::Generics` at the call
+            // sites we invoke `IsSuggestable::is_suggestable`.
```

Comment:
Is this comment revelant? I wrote this comment, but in somewhat different context?

Reply:
It is the same mechanism: inspect the Symbol to see if it  starts with `impl `. We have the same situation here. Maybe I can add a method or modify `Ty::is_impl_trait` to also do that check instead.

Reply:
Yeah, but why are we looking at arbitrary generics? The reason that this hack is not necessary in the rest of this file is because we're only looking at generics that come from specific predicates that we know have failed. That's why below I've said that this logic is pretty different from all of the other parameter searching logic, and why I'm confused.

Reply:
I'll rebase on top of yours, as there are a couple of cases that are still not fully handled in yours.